stl/inc/xutility

```diff
@@ -4198,7 +4198,383 @@ _NODISCARD _CONSTEXPR17 move_iterator<_Iter> make_move_iterator(_Iter _It) { //
+    }
+
+    // clang-format off
+    template <class _Iter2, class _Se2>
+        requires convertible_to<const _Iter2&, _Iter> && convertible_to<const _Se2&, _Se>
+              && assignable_from<_Iter&, const _Iter2&> && assignable_from<_Se&, const _Se2&>
+    common_iterator& operator=(const common_iterator<_Iter2, _Se2>& _Right) noexcept(
+        is_nothrow_assignable_v<_Iter, const _Iter2&>&& is_nothrow_assignable_v<_Se, const _Se2&>&&
+            is_nothrow_constructible_v<_Iter, const _Iter2&>&& is_nothrow_constructible_v<_Se, const _Se2&>&&
+                is_nothrow_destructible_v<_Iter>&& is_nothrow_destructible_v<_Se>) // strengthened
+    // clang-format on
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Right._Mystate != _State::_Nonthing, "common_iterator can only be assigned if it holds a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter) {
+            if constexpr (!is_nothrow_assignable_v<_Iter, const _Iter2&>) {
+                _Mystate = _State::_Nothing;
+            }
+            _Myiter = _Right._Myiter;
+            if constexpr (!is_nothrow_assignable_v<_Iter, const _Iter2&>) {
+                _Mystate = _State::_Holds_iter;
+            }
+        } else if (_Mystate == _State::_Holds_sentinel && _Right._Mystate == _State::_Holds_sentinel) {
+            if constexpr (!is_nothrow_assignable_v<_Se, const _Se2&>) {
+                _Mystate = _State::_Nothing;
+            }
+            _Mysentinel = _Right._Mysentinel;
+            if constexpr (!is_nothrow_assignable_v<_Se, const _Se2&>) {
+                _Mystate = _State::_Holds_sentinel;
+            }
+        } else if (_Right._Mystate == _State::_Holds_iter) {
+            if (_Mystate == _State::_Holds_sentinel) {
+                if constexpr (!is_nothrow_destructible_v<_Se>) {
+                    _Mystate = _State::_Nothing;
+                }
+                _Mysentinel.~_Se();
+            }
+            if constexpr (!is_nothrow_constructible_v<_Iter, const _Iter2&>) {
+                _Mystate = _State::_Nothing;
+            }
+            _Construct_in_place(_Myiter, _Right._Myiter);
+            _Mystate = _State::_Holds_iter;
+        } else if (_Right._Mystate == _State::_Holds_sentinel) {
+            if (_Mystate == _State::_Holds_iter) {
+                if constexpr (!is_nothrow_destructible_v<_Iter>) {
+                    _Mystate = _State::_Nothing;
+                }
+                _Myiter.~_Iter();
+            }
+            if constexpr (!is_nothrow_constructible_v<_Se, const _Se2&>) {
+                _Mystate = _State::_Nothing;
+            }
+            _Construct_in_place(_Mysentinel, _Right._Mysentinel);
+            _Mystate = _State::_Holds_sentinel;
+        }
+        return *this;
+    }
+
+    ~common_iterator() noexcept(is_nothrow_destructible_v<_Iter>&& is_nothrow_destructible_v<_Se>) // strengthened

```

Comment:
I think this should be unconditionally `noexcept` and not "strengthened". The STL is allowed to assume that user destructors never emit exceptions (even if they happen to be marked `noexcept(false)`) yet the STL is required to behave as if all of its destructors are `noexcept` which implies that they actually need to be marked `noexcept` to prevent them from being infected by `noexcept(false)` user types.

We may not be perfect about this discipline (I recall going back and forth on dtor markings in the past) but I observe that we have `~vector() noexcept`.

Reply:
I assume that this also applies for the destruction in `operator=`

Reply:
Ditto "we don't have to assume because `Destructible` actually enforces that these destructions are `noexcept`."